runtime/src/accounts_db.rs

```diff
@@ -2876,59 +2975,102 @@ impl AccountsDB {
                 .iter()
-                .map(|iter_item| {
+                .filter_map(|iter_item| {
                     let key = iter_item.key();
                     let account = &iter_item.value().account;
-                    let hash = iter_item.value().hash;
-                    total_size += (account.data.len() + STORE_META_OVERHEAD) as u64;
-                    ((key, account), hash)
+                    if should_flush_f
+                        .as_mut()
+                        .map(|should_flush_f| should_flush_f(key, account))
+                        .unwrap_or(true)
+                    {
+                        let hash = iter_item.value().hash;
+                        total_size += (account.data.len() + STORE_META_OVERHEAD) as u64;
+                        // If any entry is persisted for this slot, it's not dead.
+                        // Mark this down so we don't remove pertinent information like
+                        // bank hash which is needed for snapshotting.
+                        is_dead_slot = false;
+                        Some(((key, account), hash))
+                    } else {
+                        // If we don't flush, we have to remove the entry from the
+                        // index, since it's equivalent to purging
+                        purged_slot_pubkeys.insert((slot, *key));
+                        pubkey_to_slot_set.push((*key, slot));
+                        None
+                    }
                 })
                 .unzip();
-            let aligned_total_size = self.page_align(total_size);
-
-            // This ensures that all updates are written to an AppendVec, before any
-            // updates to the index happen, so anybody that sees a real entry in the index,
-            // will be able to find the account in storage
-            let flushed_store =
-                self.create_and_insert_store(slot, aligned_total_size, "flush_slot_cache");
-            self.store_accounts_custom(
-                slot,
-                &accounts,
-                &hashes,
-                Some(Box::new(move |_, _| flushed_store.clone())),
-                None,
-                false,
-            );
-            // If the above sizing function is correct, just one AppendVec is enough to hold
-            // all the data for the slot
-            assert_eq!(
-                self.storage
-                    .get_slot_stores(slot)
-                    .unwrap()
-                    .read()
-                    .unwrap()
-                    .len(),
-                1
-            );
+
+            // Remove the account index entries from earlier roots that are outdated by later roots.
+            // Safe because queries to the index will be reading updates from later roots.
+            self.purge_slot_cache_keys(slot, purged_slot_pubkeys, pubkey_to_slot_set, is_dead_slot);
+
+            if !is_dead_slot {
```

Comment:
maybe this flag (`is_dead_slot`) is easily computable from `accounts.is_empty()`?

Reply:
ah yeah, that's better!

Reply:
:eyes: